example/graph/README.md

```diff
@@ -0,0 +1,119 @@
+the option `build-all` or `build-example` to the make command.
+
+### Usage
+Once the examples are built, the binaries will be located in the `example` subdirectory of your 
+build directory. The D2GC example application will be in `KokkosKernels_Example_Distance2GraphColor.exe`.
+Executing the application with no parameters will print out the usage options.
+
+```
+Usage:
+  ./KokkosKernels_Example_Distance2GraphColor.exe [parameters]
+
+Parameters:
+  Parallelism (select one of the following):
+      --serial <N>        Execute serially.
+      --threads <N>       Use N posix threads.
+      --openmp <N>        Use OpenMP with N threads.
+      --cuda              Use CUDA
+
+  Required Parameters:
+      --amtx <filename>   Input file in Matrix Market format (.mtx).
+
+      --algorithm <algorithm_name>   Set the algorithm to use.  Allowable values are:
+                 COLORING_D2_MATRIX_SQUARED  - Matrix-squared + Distance-1 method.
+                 COLORING_D2_SERIAL          - Serial algorithm (must use with 'serial' mode)
+                 COLORING_D2_VB              - Vertex Based method using boolean forbidden array (Default).
+                 COLORING_D2_VB_BIT          - VB with Bitvector Forbidden Array
+                 COLORING_D2_VB_BIT_EF       - VB_BIT with Edge Filtering
+
+  Optional Parameters:
+      --output-histogram              Print out a histogram of the colors.
+      --output-graphviz               Write the output to a graphviz file (G.dot).
+                                      Note: Vertices with color 0 will be filled in and colored
+      --output-graphviz-vert-max <N>  Upper limit of vertices in G to allow graphviz output. Default=1500.
+                                      Requires --output-graphviz to also be enabled.
+      --validate                      Check that the coloring is a valid distance-2 graph coloring
+      --verbose-level <N>             Set verbosity level [0..5] where N > 0 means print verbose messags.
+                                      Default: 0
+      --help                          Print out command line help.
+```
+
+
+### Example Use
+
+Running the example with the following parameters:
+```
+./KokkosKernels_Example_Distance2GraphColor.exe \
+        --openmp 8 \
+        --algorithm COLORING_D2_VB_BIT \
+        --output-histogram \
+        --validate \
+        --amtx ${HOME}/Data/MatrixMarket/vanHeukelum/cage11.mtx
+```
+runs the Vertex-Based + BitVector based forbidden-array variation using 8 OpenMP threads
+on the [vanHeukelum][1] [Cage11][2] graph. The coloring will be checked for correctness and a histogram 
+summary will be printed out. This results in the following output:
+```
+Run Graph Color D2 (COLORING_D2_VB_BIT)
+
+Distance-2 Graph Coloring is VALID
+
+Distance-2 Color Histogram:
+0 1767 1737 1774 1748 1619 1610 1596 1494 1467 1395 1358 1288 1272 1188 1161 1082 1050 1016 962 ... ... ... 30 26 22 21 16 18 13 13 12 8 8 7 6 5 5 4 3 2 2 2
```

Comment:
@william76 Does the histogram dump start at color 0? It would be nice to start at color 1 since no vertex will ever be color 0. Maybe just print out a subview from 1 to the end.

Reply:
@william76 Everything else is good.

Reply:
@brian-kelley How do I extract a sub-view for elements 1..N-1 of a view?

Reply:
@william76 This would be from 1 to the end, so ``Kokkos::subview(histogram, std::make_pair(1, histogram.extent(0)))``.

Reply:
@brian-kelley hmmm, I changed my print function to do this:
```C++
auto histogram_slice = Kokkos::subview(histogram, std::make_pair(1,histogram.extent(0)));
```
but this doesn't compile:
```
In file included from /Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_View.hpp:490:0,
                 from /Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_Parallel.hpp:52,
                 from /Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_Serial.hpp:55,
                 from /Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_Core.hpp:53,
                 from /Users/wcmclen/dev/kk-dev/source/KokkosKernels/example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:51:
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/impl/Kokkos_ViewMapping.hpp: In instantiation of ‘struct Kokkos::Impl::is_integral_extent<0, std::pair<int, long unsigned int> >’:
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/impl/Kokkos_ViewMapping.hpp:3231:12:   required from ‘class Kokkos::Impl::ViewMapping<void, Kokkos::ViewTraits<int*, Kokkos::HostSpace>, std::pair<int, long unsigned int> >’
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_View.hpp:2487:1:   required by substitution of ‘template<class D, class ... P, class ... Args> typename Kokkos::Impl::ViewMapping<void, Kokkos::ViewTraits<DataType, Properties ...>, Args ...>::type Kokkos::subview(const Kokkos::View<D, P ...>&, Args ...) [with D = int*; P = {Kokkos::HostSpace}; Args = {std::pair<int, long unsigned int>}]’
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/install/include/KokkosGraph_Distance2Color_impl.hpp:527:39:   required from ‘void KokkosGraph::Impl::GraphColorDistance2<HandleType, lno_row_view_t_, lno_nnz_view_t_, clno_row_view_t_, clno_nnz_view_t_>::print_color_histogram() [with HandleType = KokkosGraph::GraphColorDistance2Handle<const int, const int, const int, Kokkos::OpenMP, Kokkos::HostSpace, Kokkos::HostSpace>; lno_row_view_t_ = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; lno_nnz_view_t_ = Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; clno_row_view_t_ = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; clno_nnz_view_t_ = Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >]’
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/install/include/KokkosGraph_Distance2Color_impl.hpp:1795:9:   required from ‘void KokkosGraph::Impl::graph_print_distance2_color_histogram(KernelHandle*, typename KernelHandle::nnz_lno_t, typename KernelHandle::nnz_lno_t, lno_row_view_t_, lno_nnz_view_t_, lno_col_view_t_, lno_colnnz_view_t_, bool) [with KernelHandle = KokkosKernels::Experimental::KokkosKernelsHandle<int, int, double, Kokkos::OpenMP, Kokkos::HostSpace, Kokkos::HostSpace>; lno_row_view_t_ = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; lno_nnz_view_t_ = Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; lno_col_view_t_ = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; lno_colnnz_view_t_ = Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; typename KernelHandle::nnz_lno_t = int]’
/Users/wcmclen/dev/kk-dev/source/KokkosKernels/example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:454:65:   required from ‘void KokkosKernels::Example::run_example(CrsGraph_type, DataType, KokkosKernels::Example::Parameters) [with ExecSpace = Kokkos::OpenMP; DataType = int; CrsGraph_type = Kokkos::StaticCrsGraph<int, Kokkos::LayoutLeft, Kokkos::OpenMP, int, void>; TempMemSpace = Kokkos::HostSpace; PersistentMemSpace = Kokkos::HostSpace]’
/Users/wcmclen/dev/kk-dev/source/KokkosKernels/example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:505:9:   required from ‘void KokkosKernels::Example::driver(KokkosKernels::Example::Parameters) [with size_type = int; lno_type = int; exec_space = Kokkos::OpenMP; hbm_mem_space = Kokkos::HostSpace]’
/Users/wcmclen/dev/kk-dev/source/KokkosKernels/example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:544:119:   required from here
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/impl/Kokkos_ViewMapping.hpp:315:3: error: static assertion failed: subview argument must be either integral or integral extent
   static_assert( value ||
   ^~~~~~~~~~~~~
```
Any of that make sense to you?

Reply:
@william76 I think Kokkos is confused because the elements of the pair are different. Could you try casting the 1 to ``size_type``? I think that'll work...

Reply:
@brian-kelley well, that worked.  :man_facepalming:
```C++
auto histogram_slice = Kokkos::subview(histogram, std::make_pair((size_t)1, histogram.extent(0)));
```
It looks like it does the right thing. I'll do a quick rebuild and if all looks ok I'll push the change up.

Reply:
@brian-kelley I pushed the change in.